source/dataman/CLS/CLSDbUpgrade1Pt3.cpp

```diff
@@ -0,0 +1,134 @@
+			databaseToUpgrade_->rollbackTransaction();
+			AMErrorMon::alert(this, CLSDBUPGRADE1PT3_COULD_NOT_CREATE_TABLE, QString("Could not create %1.").arg(clsConfigurationTableName));
+			return false;
+		}
+
+		QString configurationROITableName = QString("%1_regionsOfInterest").arg(clsConfigurationTableName);
+		QStringList ROITableColumnNames = QStringList() << "id1" << "table1" << "id2" << "table2";
+		QStringList ROITableColumnTypes = QStringList() << "INTEGER" << "TEXT" << "INTEGER" << "TEXT";
+
+		if (!databaseToUpgrade_->ensureTable(configurationROITableName, ROITableColumnNames, ROITableColumnTypes)) {
+			databaseToUpgrade_->rollbackTransaction();
+			AMErrorMon::alert(this, CLSDBUPGRADE1PT3_COULD_NOT_CREATE_TABLE, QString("Could not create %1.").arg(configurationROITableName));
+			return false;
+		}
+
+		QStringList indexColumns = QStringList() << "id1" << "id2" ;
+		foreach (QString column, indexColumns) {
+			if (!databaseToUpgrade_->createIndex(configurationROITableName, column)) {
+				databaseToUpgrade_->rollbackTransaction();
+				AMErrorMon::alert(this, CLSDBUPGRADE1PT3_COULD_NOT_CREATE_INDEX, QString("Could not create index for %1 on %2.").arg(configurationROITableName).arg(column));
+				return false;
+			}
+		}
+	}
+
+	// Stage 2:  update the inheritance of the beamline UserConfiguration
+	///////////////////////////////////////////////////
+
+	QString beamlineUserConfigurationClassNameTemplate = "%1UserConfiguration";
+	QString tableDescriptionTemplate = "%1 User Configuration Database Object";
+
+	QStringList dboColumnNames = QStringList() << "AMDbObjectType" << "tableName" << "description" << "version" << "inheritance";
+	QVariantList dboValues ;
+	QString beamlineUserConfigurationClassName;
+
+	// ------ find the beamline information and update the inheritence information ------
+	QStringList beamlinesToCheck = QStringList() << "BioXAS" << "IDEAS" << "PGM" << "SGM" << "SXRMB" << "VESPERS"; //no REIXS, BioXASImaging
+
+	foreach (QString beamlineName, beamlinesToCheck) {
+		beamlineUserConfigurationClassName = beamlineUserConfigurationClassNameTemplate.arg(beamlineName);
+
+		QVariantList dboClassIds = databaseToUpgrade_->retrieve("AMDbObjectTypes_table", "id");
+		QVariantList dboClassNames = databaseToUpgrade_->retrieve("AMDbObjectTypes_table", "AMDbObjectType");
+		int index = dboClassNames.indexOf(beamlineUserConfigurationClassName);
+		if ( index >= 0) {
+
+			dboValues = QVariantList() << beamlineUserConfigurationClassName << QString("%1_table").arg(beamlineUserConfigurationClassName) << tableDescriptionTemplate.arg(beamlineName)
+									   << 1 << QString("%1;%2").arg(beamlineUserConfigurationClassName).arg(clsUserConfigurationClassInheritence);
+
+			int dbResult = databaseToUpgrade_->insertOrUpdate(dboClassIds.at(index).toInt(), "AMDbObjectTypes_table", dboColumnNames, dboValues);
```

Comment:
If this is inserting to the AMDbObjectTypes_table then you need to make sure that the AMDbObjectTypes_allColumns, AMDbObjectTypes_visibleColumns, and AMDbObjectTypes_loadColumns are properly populated.